swift/looker/rtl/oauthSession.swift

```diff
@@ -0,0 +1,157 @@
+    var hexStr: String {
+        bytes.map { String(format: "%02x", $0) }.joined()
+    }
+}
+
+// from https://stackoverflow.com/a/40089462/74137
+extension Data {
+    struct HexEncodingOptions: OptionSet {
+        let rawValue: Int
+        static let upperCase = HexEncodingOptions(rawValue: 1 << 0)
+    }
+
+    func hexStr(_ options: HexEncodingOptions = []) -> String {
+        let hexDigits = Array((options.contains(.upperCase) ? "0123456789ABCDEF" : "0123456789abcdef").utf16)
+        var chars: [unichar] = []
+        chars.reserveCapacity(2 * count)
+        for byte in self {
+            chars.append(hexDigits[Int(byte / 16)])
+            chars.append(hexDigits[Int(byte % 16)])
+        }
+        return String(utf16CodeUnits: chars, count: chars.count)
+    }
+}
+
+@available(OSX 10.15, *)
+class OAuthSession: AuthSession {
+    var code_verifier: String = ""
+
+    override init(_ settings: IApiSettings, _ transport: ITransport? = nil) {
+        super.init(settings, transport)
+    }
+
+    func requestToken(_ body: Values) throws -> AuthToken {
+        let response : SDKResponse<AccessToken, SDKError> = self.transport.request(
+            HttpMethod.POST,
+            "/api/token",
+            nil,
+            // This body needs to be www-form-urlencoded for the login method
+            // encoding it to String here makes that happen automatically
+            encodeParams(body),
+            nil,
+            nil
+        )
+        do {
+            let token = try self.ok(response)
+            return self._authToken.setToken(token)
+        } catch {
+            throw error
+        }
+    }
+
+    override func getToken() throws -> AuthToken {
+        if (!self.isAuthenticated()) {
+            if (!self.activeToken.refresh_token.isEmpty) {
+                let config = self.settings.readConfig(nil)
+                // fetch the token
+                let _ = try self.requestToken([
+                    "grant_type": "refresh_token",
+                    "refresh_token": self.activeToken.refresh_token,
+                    "client_id": config["client_id"],
+                    "redirect_uri": config["redirect_uri"]
+                ])
+            }
+        }
+        return self.activeToken
+    }
+
+    /*
+     Generate an OAuth2 authCode request URL
+     */
+    func createAuthCodeRequestUrl(scope: String, state: String) throws -> String {
+        self.code_verifier = self.sha256Hash(try! self.secureRandom(32))
+        let config = self.settings.readConfig(nil)
+        let looker_url = config["looker_url"]!
+        let url = addQueryParams("\(looker_url)/auth",
+            [
+                "response_type": "code",
+             "client_id": config["client_id"],
+            "redirect_uri": config["redirect_uri"],
+            "scope": scope,
+            "state": state,
+            "code_challenge_method": "S256",
+            "code_challenge": self.code_verifier
+        ])
+        return url
```

Comment:
Looks like same issue here:
```suggestion
        self.code_verifier = try! self.secureRandom(32)
        let code_challenge = self.sha256Hash(self.code_verifier)
        let config = self.settings.readConfig(nil)
        let looker_url = config["looker_url"]!
        let url = addQueryParams("\(looker_url)/auth", [
            "response_type": "code",
            "client_id": config["client_id"],
            "redirect_uri": config["redirect_uri"],
            "scope": scope,
            "state": state,
            "code_challenge_method": "S256",
            "code_challenge": code_challenge
        ])
        return url
```